nav2_smac_planner/src/node_hybrid.cpp

```diff
@@ -402,72 +415,89 @@ void NodeHybrid::resetObstacleHeuristic(
-  // search-planning request that dynamically updates during search as needed.
   const int size_x_int = static_cast<int>(size_x);
   const unsigned int size_y = sampled_costmap->getSizeInCellsY();
-  const float sqrt_2 = sqrt(2);
-  unsigned int mx, my, mx_idx, my_idx;
-  unsigned int idx = 0, new_idx = 0;
-  float last_accumulated_cost = 0.0, travel_cost = 0.0;
-  float heuristic_cost = 0.0, current_accumulated_cost = 0.0;
-  float cost = 0.0, existing_cost = 0.0;
 
   const std::vector<int> neighborhood = {1, -1,  // left right
     size_x_int, -size_x_int,  // up down
     size_x_int + 1, size_x_int - 1,  // upper diagonals
     -size_x_int + 1, -size_x_int - 1};  // lower diagonals
 
-  while (!obstacle_heuristic_queue.empty()) {
-    // get next one
-    idx = obstacle_heuristic_queue.front();
-    obstacle_heuristic_queue.pop();
-    last_accumulated_cost = obstacle_heuristic_lookup_table[idx];
+  // start_x and start_y have changed since last call
+  // we need to recompute 2D distance heuristic and reprioritize queue
+  for (auto & n : obstacle_heuristic_queue) {
+    n.first = -obstacle_heuristic_lookup_table[n.second] +
+      distanceHeuristic2D(n.second, size_x, start_x, start_y);
+  }
+  std::make_heap(
+    obstacle_heuristic_queue.begin(), obstacle_heuristic_queue.end(),
+    ObstacleHeuristicComparator{});
 
+  float c_cost, cost, travel_cost, new_cost, existing_cost;
+  unsigned int mx, my, mx_idx, my_idx;
+  unsigned int new_idx = 0;
+  const float sqrt_2 = sqrt(2);
 
-    if (idx == start_index) {
-      // costs are doubled due to downsampling
-      return 2.0 * last_accumulated_cost;
+  while (!obstacle_heuristic_queue.empty()) {
+    ObstacleHeuristicElement c = obstacle_heuristic_queue.front();
+    std::pop_heap(
+      obstacle_heuristic_queue.begin(), obstacle_heuristic_queue.end(),
+      ObstacleHeuristicComparator{});
+    obstacle_heuristic_queue.pop_back();
+    c_cost = obstacle_heuristic_lookup_table[c.second];
+    if (c_cost > 0.0f) {
+      // cell has been processed and closed, no further cost improvements
+      // are mathematically possible thanks to euclidean distance heuristic consistency
+      continue;
     }
+    c_cost = -c_cost;
+    obstacle_heuristic_lookup_table[c.second] = c_cost;  // set a positive value to close the cell
 
-    my_idx = idx / size_x;
-    mx_idx = idx - (my_idx * size_x);
+    my_idx = c.second / size_x;
+    mx_idx = c.second - (my_idx * size_x);
 
     // find neighbors
     for (unsigned int i = 0; i != neighborhood.size(); i++) {
-      new_idx = static_cast<unsigned int>(static_cast<int>(idx) + neighborhood[i]);
-      cost = static_cast<float>(sampled_costmap->getCost(idx));
-      travel_cost =
-        ((i <= 3) ? 1.0 : sqrt_2) + (((i <= 3) ? 1.0 : sqrt_2) * cost_penalty * cost / 252.0);
-      current_accumulated_cost = last_accumulated_cost + travel_cost;
+      new_idx = static_cast<unsigned int>(static_cast<int>(c.second) + neighborhood[i]);
```

Comment:
@SteveMacenski @afrixs new_idx is int so if the result of static_cast<unsigned int>() exceeds the max of int, new_idx will be negative and the cost lookup segfaults.

In may case:
 c.second = 135
 neighborhood[i] = -407
 this leads to new_idx = -272.
 The cast in getCost to unsigned int makes the lookup at an array index well out of bounds. -> Segfault

Reply:
@mangoschorle Did you check this with my latest updates from yesterday? I moved the `cost = static_cast<float>(sampled_costmap->getCost(new_idx));` line into the `if (new_idx > 0 && new_idx < size_x * size_y)` block and I see no reason here for Segfaults any more. Also `new_idx` is `unsigned int`, not `int`.

Reply:
@afrixs My bad. I hadn't looked at the new version.
Just a question:
The neighbourhood doesn't work on costmap edges, or am I wrong. The +1 might actually not be the cell to the right but in the next line on the very left?

Reply:
There's a check for the edges, see [these lines](https://github.com/ros-planning/navigation2/blob/85cf934de1c0cf8d335f8d46440a96636410b8ee/nav2_smac_planner/src/node_hybrid.cpp#L504)

Reply:
This is the only remaining open comment @mangoschorle is this solved to your satisfaction to "resolve" it?

Reply:
Its good! I have zero experience with github. Do I need to press some "resolved" button, or are you asking because you didn't see any reply from my side?

Reply:
Usually, yeah the "Resolve conversation" button. Don't worry I got it, just wanted to make sure this was fully settled.